### PR TITLE
Skip files that are not changed from sync

### PR DESCRIPTION
**Pull request by Fatme**
This will resolve the problem with endless loop when the application has before-prepare hook that modifies some files from app folder.

### CODE REVIEW

File: services/hash-service.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+			future.wait();
+			return cryptoHash.digest(hashEncoding);
+		}).future<string>()();
+	}
+
+	private validateInputParameters(filePath: string, inputEncoding: string, hashAlgorithm: string, hashEncoding: string): IFuture<void> {
+		return (() => {
+			if(!this.$fs.exists(filePath).wait()) {
+				this.$errors.fail(util.format("Specified file %s does not exist.", filePath));
```
Comment (teobugslayer) [inline services/hash-service.ts]: Consider template strings (with backquotes)

File: services/hash-service.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+			}
+
+			if(!_.contains(HashService.validInputEncodings, inputEncoding)) {
+				this.$errors.fail(util.format("Specified input file encoding %s is not valid. Valid values are %s", inputEncoding, HashService.validInputEncodings));
```
Comment (teobugslayer) [inline services/hash-service.ts]: Again, consider back quotes

File: services/hash-service.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			if(!_.contains(HashService.validHashEncodings, hashEncoding)) {
+				this.$errors.fail(util.format("Specified hash encoding %s is not valid. Valid values are %s", hashEncoding, HashService.validHashEncodings));
+			}
+		}).future<void>()();
+	}
+
+	private validateHashAlgorithm(hashAlgorithm: string): void {
+		let hashes = crypto.getHashes();
+		if(!_.contains(hashes, hashAlgorithm)) {
+			this.$errors.fail(util.format("Specified hash algorithm %s is not valid. Valid algorithms are %s", hashAlgorithm, hashes));
```
Comment (teobugslayer) [inline services/hash-service.ts]: Is this fail or failWithoutHelp?

File: services/hash-service.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+			}
+
+			if(!_.contains(HashService.validInputEncodings, inputEncoding)) {
+				this.$errors.fail(util.format("Specified input file encoding %s is not valid. Valid values are %s", inputEncoding, HashService.validInputEncodings));
+			}
+
+			this.validateHashAlgorithm(hashAlgorithm);
+
+			if(!_.contains(HashService.validHashEncodings, hashEncoding)) {
+				this.$errors.fail(util.format("Specified hash encoding %s is not valid. Valid values are %s", hashEncoding, HashService.validHashEncodings));
```
Comment (teobugslayer) [inline services/hash-service.ts]: I think failWithoutHelp is more appropriate here

File: services/usb-livesync-service-base.ts [hunk 5]
```diff
@@ -108,24 +111,40 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
-								let canExecuteFastLiveSync = data.canExecuteFastLiveSync && data.canExecuteFastLiveSync(filePath);
-
-								if (synciOSSimulator && !canExecuteFastLiveSync) {
-									that.batchSimulatorLiveSync(data, filePath);
+							that.$dispatcher.dispatch(() => (() => {
+								let fileHash = that.$fs.getFileShasum(filePath).wait();
+								if (fileHash === that.fileHashes[filePath]) {
+									that.$logger.trace(`Skipping livesync for ${filePath} file with ${fileHash} hash.`);
+									return;
+								} else {
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: Is the else operator needed? The if ends with an `return`.

File: services/usb-livesync-service-base.ts [hunk 6]
```diff
@@ -108,24 +111,40 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
 					this.on('all', (event: string, filePath: string) => {
 						if (event === "added" || event === "changed") {
-							if (!that.isFileExcluded(filePath, data.excludedProjectDirsAndFiles, data.projectFilesPath)) {
-								let canExecuteFastLiveSync = data.canExecuteFastLiveSync && data.canExecuteFastLiveSync(filePath);
-
-								if (synciOSSimulator && !canExecuteFastLiveSync) {
-									that.batchSimulatorLiveSync(data, filePath);
+							that.$dispatcher.dispatch(() => (() => {
+								let fileHash = that.$fs.getFileShasum(filePath).wait();
+								if (fileHash === that.fileHashes[filePath]) {
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: I think the hash check must happen after the isFileExcluded check

File: services/usb-livesync-service-base.ts [hunk 7]
```diff
@@ -108,24 +111,40 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+									}
 								}
-							}
+
+							}).future<void>()());
 						}
 
 						if (event === "deleted") {
+							if (that.fileHashes[filePath]) {
+								delete that.fileHashes[filePath];
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: `delete` prevents V8 from optimizing this function, making it 10-20 times slower. Can we avoid it? It is an optional improvement, and I do not insist on it, but if we can do it without making the code too ugly, it is good idea to do it.

**Comment by teobugslayer** [comment]
Most excellent, :+1:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Pull request merged**